sqlcapture/capture.go

```diff
@@ -402,21 +448,29 @@ func (c *Capture) updateState(ctx context.Context) error {
 
 	// Likewise streams may be removed from the catalog, and we need to forget
 	// the corresponding state information.
-	for streamID, state := range c.State.Streams {
-		var _, streamExistsInCatalog = c.Bindings[streamID]
-		if !streamExistsInCatalog && state.Mode != TableModeIgnore {
-			logrus.WithField("stream", streamID).Info("stream removed from catalog")
-			c.State.Streams[streamID] = &TableState{Mode: TableModeIgnore, dirty: true}
+	streamExistsInCatalog := func(sk boilerplate.StateKey) bool {
+		for _, b := range c.Bindings {
+			if b.StateKey == sk {
+				return true
+			}
+		}
+		return false
+	}
+	for stateKey, state := range c.State.Streams {
+		if !streamExistsInCatalog(stateKey) && state.Mode != TableModeIgnore {
+			logrus.WithField("stateKey", stateKey).Info("stream removed from catalog")
+			c.State.Streams[stateKey] = &TableState{Mode: TableModeIgnore, dirty: true}
 		}
 	}
 
-	// If no bindings are present, forget the old replication cursor. This is safe
-	// because logically if no streams are active then we can't miss any events of
-	// interest when the replication stream jumps ahead, and doing this allows the
-	// user an easy recovery path after WAL deletion, they just need to disable all
-	// bindings and then re-enable them (which will cause them all to get backfilled
-	// anew, as they should after such an event).
-	if len(c.Bindings) == 0 {
+	// If all bindings have been removed or had their backfill counter incremented, forget the old
+	// replication cursor. This is safe because logically if no streams are active or a re-backfill
+	// has been requested for them then we can't miss any events of interest when the replication
+	// stream jumps ahead, and doing this allows the user an easy recovery path after WAL deletion,
+	// they just need to either increment the backfill counter for all bindings, or disable all
+	// bindings and then re-enable them (which will cause them all to get backfilled anew, as they
+	// should after such an event).
+	if allStreamsAreNew {
```

Comment:
Hmm, so changing the check to work this way means that if the backfill counters for every binding are incremented all at once we'll also reset the cursor? Nice!